File: files/jvm/src/main/java/com/swoval/files/RegisterableWatchService.java
package com.swoval.files;

import java.io.IOException;
import java.nio.file.Path;
import java.nio.file.WatchEvent;
import java.nio.file.WatchKey;

/**
 * Augments the <a
 * href="https://docs.oracle.com/javase/7/docs/api/java/nio/file/WatchService.html">java.nio.file.WatchService</a>
 * with a {@link RegisterableWatchService#register} method. This is because <a
 * href="https://docs.oracle.com/javase/7/docs/api/java/nio/file/Path.html#register(java.nio.file.WatchService,%20java.nio.file.WatchEvent.Kind...)">Path.register</a>
 * does not work on custom watch services.
 */
public interface RegisterableWatchService extends java.nio.file.WatchService {

  /**
   * Register a path for monitoring.
   *
   * @param path The path to monitor.
   * @param kinds The types of events to monitor.
   * @return WatchKey the key returned by the WatchService
   * @throws IOException when the path cannot be registered
   */
  WatchKey register(final Path path, final WatchEvent.Kind<?>... kinds) throws IOException;
}

File: files/js/src/main/scala/com/swoval/files/RegisterableWatchService.scala
package com.swoval.files

import java.nio.file.{ Path, WatchEvent, WatchKey, WatchService => JWatchService }
import java.util.concurrent.TimeUnit

private[files] object RegisterableWatchService {
  def newWatchService(): Registerable = new Registerable {
    override def register(path: Path, kinds: WatchEvent.Kind[_]*): WatchKey = null
    override def close(): Unit = {}
    override def poll(): WatchKey = null
    override def poll(timeout: Long, unit: TimeUnit): WatchKey = null
    override def take(): WatchKey = null
  }
}
private[files] class RegisterableWatchService(underlying: JWatchService)
    extends JWatchService
    with Registerable {
  override def close(): Unit = {}
  override def poll(): WatchKey = null
  override def poll(timeout: Long, unit: TimeUnit): WatchKey = null
  override def take(): WatchKey = null
  override def register(path: Path, kinds: WatchEvent.Kind[_]*): WatchKey = null
}
